[하티] 3, 4단계 영화 극장 선택 제출합니다.

sujin9 opened the pull request:
안녕하세요, 커트! :)
주말은 잘 보내셨나요?

우선, 지난 1, 2 단계 때 주신 피드백들에 감사드려요!
3단계를 시작하기에 앞서 우선 주신 피드백들을 적용하려고 해보았습니다. :)

그 이후 3, 4 단계를 진행하였는데요, 시간 상의 부족함이 있어서 모든 presenter 로직에 대한 테스트는 작성하지 못했습니다 😢 
또한, 현재 코드에서 
- Presenter가 값을 반환하는 형태가 아닌 view의 함수를 호출하는 방식으로 리팩토링하기
- repository interface들은 역할이 domain에 속하는 것이 맞다고 생각되어서 도메인 모듈로 옮겨 리팩토링하기
- 버튼 클릭 등의 이벤트를 바인딩 어댑터를 이용하여 처리하도록 리팩토링하기

등으로 개선해보고 싶은 마음이 있는데, 반영하지 못해 아쉬운 마음이에요 :( 
이 부분은 리뷰 요청을 보낸 이후에 마저 개선해보겠습니다!

또한, 미션을 진행하는 과정에서 궁금한 점이 있는데요,
- view와 presenter를 구분하는 기준이 무엇일지 많은 고민이 되었는데요, 현재 제 코드가 MVP 패턴에 맞게 잘 분리가 되어있는지 궁금합니다!
- 또한 presenter에서 무엇을 테스트 해야할지 잘 감이 잡히지 않았고, 또한 presenter를 테스트하면서 그러면 ui test에서는 무엇을 테스트해야하지? 하는 생각들이 들었어요. 이 부분에 대해서 커트께서 갖고 계신 생각이나 기준이 있으시다면 궁금해요.
- view에서 presenter, presenter에서 view의 함수를 호출하려면 override 메서드가 되어야 하는데, 이 과정에서 함수들이 private하지 않게 되는 것에 대해서 단점은 없을지 궁금해요.

이번 미션 리뷰도 잘 부탁드립니다! 감사합니다 🌸

sujin9 committed:
fix: this가 잘못 정의된 부분 수정

- companion object 내의 this 사용 수정

sujin9 committed:
refactor: 모든 fragment가 아닌 기존의 fragment만 hide 되도록 수정

sujin9 committed:
refactor: fragment가 null인 경우에 대한 분기 처리 명시

sujin9 committed:
refactor: MovieDetailActivity MVP 패턴 적용

sujin9 committed:
refactor: 패키지 이름 수정

sujin9 committed:
test: test 삭제

- 실제 shared preference 값에 영향을 주기 때문에 바르지 못한 테스트라고 판단

sujin9 committed:
test: MovieDetailPresenter 테스트 추가

sujin9 committed:
refactor: fragment가 계속 생성되지 않도록 수정

sujin9 committed:
refactor: SeatSelectionActivity MVP 패턴 적용

sujin9 committed:
fix: 영화 예매를 한 이후 예매 내역이 바로 반영되지 않는 문제 해결

sujin9 committed:
refactor: shared preferences 관련 로직을 data로 분리, 접근을 위해 repository 적용

sujin9 committed:
refactor: SettingFragment MVP 패턴 적용

sujin9 committed:
refactor: HomeFragment MVP 패턴 적용

sujin9 committed:
feat: 예약 내역을 저장하는 데이터베이스 구현

sujin9 committed:
refactor: ReservationFragment MVP 패턴 적용

sujin9 committed:
refactor: binding 적용

sujin9 committed:
refactor: binding adapter 적용

sujin9 committed:
feat: 극장 선택 bottom sheet dialog 추가

sujin9 committed:
feat: 극장에 따라 영화 날짜 및 시간표가 다르게 나타나도록 구현

sujin9 committed:
feat: 극장 데이터 로딩을repository 를 이용하여 구현

sujin9 committed:
fix: mock data 리스트의 local time 개수가 하나일 때 오류 발생

BeokBeok commented on app/src/main/java/woowacourse/movie/data/alarm/AlarmStateDataSource.kt:
```diff
@@ -0,0 +1,21 @@
+    private val preferences = context.getSharedPreferences("movie", Context.MODE_PRIVATE)
+
+    override fun getData(): Boolean {
+        return preferences.getBoolean(SettingFragment.KEY_SWITCH, false)
```
Key값을 SettingFragment에 정의한 값을 사용하고 있네요.
AlarmStateDataSource는 어느 View에서든지 사용할 수 있어야 합니다.
SettingFragment의 의존성을 제거해보면 어떨까요?

BeokBeok commented on app/src/main/java/woowacourse/movie/ui/MainActivity.kt:
```diff
@@ -5,26 +5,27 @@ import android.view.MenuItem
-    private val bottomNavigationView: BottomNavigationView by lazy {
-        findViewById(R.id.main_bottom_navigation_view)
-    }
+    private var lastSelectedFragmentTag = ""
```
fragment의 tag를 멤버 변수로 가지고 있도록 구현해주셨네요. 👍 
다만, 멤버 변수로 fragment의 tag를 가지고 있지 않아도, fragment의 tag를 조회할 수 있어요.
lastSelectedFragmentTag 멤버 변수를 사용하지 않도록 개선해보면 어떨까요?

BeokBeok commented on app/src/main/java/woowacourse/movie/ui/home/adapter/MovieListAdapter.kt:
```diff
@@ -13,18 +15,20 @@ class MovieListAdapter(
-                MovieItemViewHolder(
-                    LayoutInflater.from(parent.context).inflate(R.layout.item_movie, parent, false),
-                ) {
+            MovieListViewType.MOVIE.value -> {
```
Adapter에서 각각 ViewHolder 관련하여 Inflate를 해주고 있네요.
Adapter와 ViewHolder의 역할에 대해서 다시 한번 학습해보시고, Inflate를 어디서 하는 것이 좋을지 고민해보면 어떨까요?

sujin9 replied:
Adapter: 데이터와 리사이클러뷰를 연결 (마치 bridge!)
ViewHolder: 리사이클러뷰에 들어가는 각각의 아이템 뷰

라고 생각합니다.

그러므로, 각각의 뷰를 그리는 부분은 view holder에서 하는 것이 맞다고 생각되어요!

BeokBeok commented on app/src/main/java/woowacourse/movie/ui/moviedetail/MovieDetailActivity.kt:
```diff
@@ -70,73 +71,132 @@ class MovieDetailActivity : AppCompatActivity() {
+        setPeopleCountView(presenter.peopleCountModel.count)
+    }
+
+    private fun setMinusButton() {
```
함수명만 봐서는 "마이너스 버튼을 설정한다" 라는 의미로 보이는데, 실제로는 마이너스버튼의 클릭리스너를 등록해주고 있네요.
함수명을 구현부와 일치시켜보면 어떨까요? (이하 관련 내용 동일)

BeokBeok commented on app/src/main/java/woowacourse/movie/ui/moviedetail/MovieDetailPresenter.kt:
```diff
@@ -0,0 +1,50 @@
+    private val theater: TheaterModel,
+) : MovieDetailContract.Presenter {
+
+    private var peopleCount = PeopleCount()
+    val peopleCountModel: PeopleCountModel
+        get() = peopleCount.toModel()
```
peopleCount를 멤버 변수로 정의해주셨네요.
멤버 변수는 클래스 내부에서 값을 편하게 제어할 수 있는 장점이 있지만, 멤버 변수의 값에 따라 함수 호출 시 결괏값을 예상할 수 없다는 단점이 있어요.
함수 호출에 의존하지 않고 테스트 코드를 작성할 수 있도록 개선해보면 어떨까요?

BeokBeok commented on app/src/main/java/woowacourse/movie/ui/reservation/ReservationFragment.kt:
```diff
@@ -4,48 +4,69 @@ import android.os.Bundle
+
+    override fun onResume() {
+        super.onResume()
+        presenter.setAdapter()
```
presenter.initAdapter와 presenter.setAdapter의 차이점은 무엇일까요?

sujin9 replied:
현재 코드 상에서는 차이점이 없습니다!
다만, initAdapter에서는 adapter를 setting 해주는 역할을,
setAdapter에서는 어댑터에 연결된 데이터를 수정해주는 부분만 수행하도록 리팩토링하려고 했습니다.

하지만, 현재 `ReservationFragment`에서는 리스트 안의 아이템이 추가/수정/삭제 등이 일어날 소지가 없다고 판단되어서, `adapter를 setting 해주는 역할`을 한 번만 수행하도록 코드를 수정하였습니다! :)

BeokBeok commented on app/src/main/java/woowacourse/movie/ui/reservation/ReservationFragment.kt:
```diff
@@ -4,48 +4,69 @@ import android.os.Bundle
+    }
+
+    private fun setTextOnEmptyState() {
+        if (presenter.isReservationEmpty()) {
+            binding.reservationEmpty.isVisible = true
         }
```
```suggestion
        binding.reservationEmpty.isVisible = presenter.isReservationEmpty()
```
코드를 간단하게 작성할 수 있어요

BeokBeok commented on app/src/main/java/woowacourse/movie/ui/seat/SeatSelectionPresenter.kt:
```diff
@@ -0,0 +1,69 @@
+    private val view: SeatSelectionContract.View,
+    private val alarmStateRepository: AlarmStateRepository,
+    private val reservationRepository: ReservationRepository,
+    val movieTitle: String,
+    val movieTime: LocalDateTime,
+    val peopleCountModel: PeopleCountModel,
```
presenter가 캡슐화하는 프로퍼티가 많아지고 있네요.
data class를 활용하여 관련 프로퍼티들을 관리해보면 어떨까요?

BeokBeok commented on app/src/main/java/woowacourse/movie/ui/setting/SettingPresenter.kt:
```diff
@@ -0,0 +1,26 @@
+        alarmStateRepository.saveData(isChecked)
+        checkSwitchState()
+
+        val tickets = reservationRepository.getData()
+        if (isChecked) {
+            tickets.forEach { view.setAlarms(it) }
+        } else {
+            tickets.forEach { view.cancelAlarms(it) }
+        }
```
```suggestion
        reservationRepository.getData()
            .forEach {
                if (isChecked) {
                    view.setAlarms(it)
                } else {
                    view.cancelAlarms(it)
                }
            }
```
forEach 중복 코드를 줄여보면 어떨까요?

BeokBeok commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,23 +1,26 @@
-</androidx.constraintlayout.widget.ConstraintLayout>
+        <androidx.fragment.app.FragmentContainerView
+            android:id="@+id/main_fragment_view"
+            android:layout_width="match_parent"
```
```suggestion
            android:layout_width="0dp"
```
ConstraintLayout의 자식뷰는 match_parent 대신 0dp를 활용해보면 어떨까요? (이하 관련 내용 동일)

BeokBeok commented on app/src/test/java/woowacourse/movie/ui/moviedetail/MovieDetailPresenterTest.kt:
```diff
@@ -0,0 +1,68 @@
+
+    @Test
+    fun PeopleCount가_1명일_때_add하면_2명이_된다() {
+        val slot = slot<Int>()
+        every { view.setPeopleCountView(capture(slot)) } answers { }
```
spy(slot)을 활용하여 테스트 코드를 작성해주셨네요. 👍 
지금도 좋지만, 객체 생성 및 주입하여 테스트 코드를 작성해보면 어떨까요? (이하 관련 내용 동일)

BeokBeok commented on app/src/test/java/woowacourse/movie/ui/setting/SettingPresenterTest.kt:
```diff
@@ -0,0 +1,56 @@
+            }
+        }
+
+        presenter = SettingPresenter(view, alarmStateRepository)
```
테스트 코드 구동 시 여기에서 에러가 발생하고 있어요.

BeokBeok commented on domain/src/main/java/woowacourse/movie/domain/seat/SeatClass.kt:
```diff
@@ -12,7 +12,7 @@ enum class SeatClass(val price: Int) {
                 in 3..4 -> S
                 5 -> A
-                else -> throw IllegalArgumentException()
+                else -> throw IllegalArgumentException("왜 여기서 오류가 나 row:$row")
```
👀

BeokBeok reviewed (changes requested):
이전 단계 피드백과 영화 극장 3,4단계 미션하시느라 고생하셨습니다. 👍 
고민해볼만한 의견들을 코맨트로 작성하였으니, 충분히 고민해보시고 도전해보세요. 💪

BeokBeok commented:
> view와 presenter를 구분하는 기준이 무엇일지 많은 고민이 되었는데요, 현재 제 코드가 MVP 패턴에 맞게 잘 분리가 되어있는지 궁금합니다!

A. 아마 MVP 패턴이 익숙지 않아서 그런 것 같아요.
전반적으로 MVP 패턴에 맞게 잘 구현해주셨습니다. 👍 
일반적으로 View는 표현만 하고, 실제 View와 관련된 데이터는 Presenter에서 관리한다고 보시면 될 것 같아요.
예를 들자면, View는 혼자서는 아무것도 못하는 바보 라고 봐주시면 이해가 편할 거에요!

> 또한 presenter에서 무엇을 테스트 해야할지 잘 감이 잡히지 않았고, 또한 presenter를 테스트하면서 그러면 ui test에서는 무엇을 테스트해야하지? 하는 생각들이 들었어요. 이 부분에 대해서 커트께서 갖고 계신 생각이나 기준이 있으시다면 궁금해요.

A. Presenter에서는 View에 보여줘야하는 데이터를 가공하는 역할과, 이 데이터를 View에 전달하는 역할을 하고 있어요.
그렇다면 Presenter에서는 아래와 같은 내용을 위주로 테스트를 진행하시면 될 거에요.
- Presenter에서 View에 보여줄 데이터로 잘 가공됐는지
- View에 전달하는 함수가 제대로 호출됐는지

> view에서 presenter, presenter에서 view의 함수를 호출하려면 override 메서드가 되어야 하는데, 이 과정에서 함수들이 private하지 않게 되는 것에 대해서 단점은 없을지 궁금해요.

A. 오히려 priavte하지 않음으로써, 대부분의 로직들을 테스트할 수 있는 장점이 있습니다.
private 함수가 많다는 것은, 그만큼 클래스가 많은 책임을 가지지 않았나 고민해보는 시점이라고 생각합니다.

sujin9 committed:
refactor: 상수가 필요한 위치에 정의되도록 수정

sujin9 committed:
refactor: presenter의 함수가 반환값을 갖지 않도록 수정

sujin9 committed:
refactor: 함수명 수정 및 코드 간결화

sujin9 committed:
refactor: 불필요한 adapter 관련 코드 삭제

sujin9 committed:
refactor: ConstraintLayout의 자식뷰에 0dp 적용

- match_parent -> 0dp

sujin9 committed:
refactor: layout을 adapter가 아닌 viewHolder에서 그리도록 수정

sujin9 committed:
refactor: 새로운 변수 사용 없이 기존 코드를 이용하여 로직이 수행되도록 수정

- `lastSelectedFragmentTag` 변수를 삭제하였다.
- 대신, 현재 bottom navigation view에서 선택되어 있는 id를 기준으로
열려있는 fragment를 구한다.

sujin9 committed:
refactor: 관련된 프로퍼티들을 하나의 데이터 클래스로 관리

- 좌석 선택 시 전달되는 영화 정보들을 MovieTicketInfo로 관리하고,
 티켓을 만들 때 이 정보와 선택한 좌석들 정보를 넘기도록 수정하였다.

sujin9 committed:
refactor: 클래스 멤버 변수를 제거하고, 함수의 인자로 값을 받도록 수정

sujin9 committed:
test: 테스트 삭제

- 로직 수정에 따라 필요 없어진 함수들 및 기능에 대한 테스트 삭제

sujin9 committed:
test: 테스트 코드 수정

sujin9 committed:
fix: 인원수가 제대로 선택되지 않는 문제 해결

sujin9 committed:
test: 좌석 선택에 관한 테스트 오류 수정 및 추가 작성

sujin9 commented:
안녕하세요, 커트!
연휴 기간 잘 보내고 계신가요 :)

다른 미션을 진행하는 데 시간이 많이 부족했다보니, 리뷰 반영이 늦어지게 되었습니다.. 🙇‍♀️ 
리뷰 요청이 늦어지게 되어 죄송하고, 다시 코드를 봐주심에 미리 감사드립니다 🙂

BeokBeok commented on app/src/main/java/woowacourse/movie/ui/moviedetail/MovieDetailActivity.kt:
```diff
@@ -48,7 +48,7 @@ class MovieDetailActivity : AppCompatActivity(), MovieDetailContract.View {
         initSpinner()
         initPeopleCountController()
-        initBookingButton(movie)
+        initBookingButtonListener(movie)
```
리스너는 init보다는 set이 더 적합해요

BeokBeok reviewed (approved):
피드백 반영하시느라 고생하셨습니다. 👍 
좋은 개발자가 되기 위한 소중한 경험이 되었길 바랍니다. 💪
